secrets/provider/kubernetes/provider.go

```diff
@@ -786,12 +823,42 @@ func (k *kubernetesClient) createClusterRole(ctx context.Context, clusterRole *r
 	return out, errors.Trace(err)
 }
 
+// updateClusterRole fetches the latest version of the specified ClusterRole,
+// replaces its Rules with those from the provided clusterRole, and updates it
+// in the cluster. This method retries on conflicts using exponential backoff
+// to handle concurrent modifications by other controllers.
+// Note that only the Rules field is updated, all other fields from the latest ClusterRole are preserved.
 func (k *kubernetesClient) updateClusterRole(ctx context.Context, clusterRole *rbacv1.ClusterRole) (*rbacv1.ClusterRole, error) {
-	out, err := k.client.RbacV1().ClusterRoles().Update(ctx, clusterRole, v1.UpdateOptions{})
-	if k8serrors.IsNotFound(err) {
-		return nil, errors.NotFoundf("cluster role %q", clusterRole.GetName())
-	}
-	return out, errors.Trace(err)
+	api := k.client.RbacV1().ClusterRoles()
+
+	var out *rbacv1.ClusterRole
+	err := retry.Call(retry.CallArgs{
+		Func: func() error {
+			patch := map[string]interface{}{
+				"rules": clusterRole.Rules,
+			}
+			data, err := json.Marshal(patch)
+			if err != nil {
+				return errors.Annotatef(err, "marshaling cluster role patch")
+			}
+			out, err = api.Patch(ctx, clusterRole.GetName(), types.StrategicMergePatchType, data, v1.PatchOptions{
+				FieldManager: resources.JujuFieldManager,
+			})
+			if k8serrors.IsNotFound(err) {
+				return errors.NotFoundf("cluster role %q", clusterRole.GetName())
+			}
+			return errors.Annotatef(err, "patching cluster role %q", clusterRole.GetName())
+		},
+		IsFatalError: func(err error) bool {
+			return !errors.Is(err, errors.NotFound) && !k8serrors.IsConflict(err)
```

Comment:
Why are we retrying on not found? Also this needs a unit test.

Reply:
So why are still retrying on not found errors? If we patch a resource, it isn't found, trying again doesn't change anything.

Reply:
hi harry, sorry missed this after resolving this comment on the unit test. I have removed the Notfound err on IsFatalError field. Thanks again for pointing this out!